### ISSUE TITLE
prune_low_magnitude retruns None

### ISSUE THREAD

**Issue by youssefsalah76** [tensorflow/model-optimization#12]
I am trying to prune a trained model , but model is saved a model_architecture.json and weights.h5 and model is loaded as follows:
with open(modeljson) as f:
            self.model = model_from_json(f.read())
        self.model.load_weights(modelfile)

But whenever I try to prune it the output from:
 new_pruned_model = sparsity.prune_low_magnitude(model, **new_pruning_params)
is always none so it crashes at the model.compile part as type none has no attribute compile
Any help please

**Comment by zhoujungis** [comment]
How do you solve it? I encounter the similar problem and could you help me?

**Comment by s36srini** [comment]
Make sure you're using Keras from Tensorflow-Python. Try this: 
`from tensorflow.python import keras` # Important: TensorFlow-Optimization uses this version of Keras. And then `from tensorflow_model_optimization.sparsity import keras as sparsity`. Your model must be built using the right version of Keras, otherwise it will return a Null model as per their source code.

**Comment by blakeedwards823** [comment]
> Make sure you're using Keras from Tensorflow-Python. Try this:
> `from tensorflow.python import keras` # Important: TensorFlow-Optimization uses this version of Keras. And then `from tensorflow_model_optimization.sparsity import keras as sparsity`. Your model must be built using the right version of Keras, otherwise it will return a Null model as per their source code. Let me know if this works, I had a similar issue a while ago.

This did not seem to work for me. What version of Keras is considered correct?

**Comment by blakeedwards823** [comment]
> I am trying to prune a trained model , but model is saved a model_architecture.json and weights.h5 and model is loaded as follows:
> with open(modeljson) as f:
> self.model = model_from_json(f.read())
> self.model.load_weights(modelfile)
> 
> But whenever I try to prune it the output from:
> new_pruned_model = sparsity.prune_low_magnitude(model, **new_pruning_params)
> is always none so it crashes at the model.compile part as type none has no attribute compile
> Any help please

Did you ever find a solution to this problem?

**Comment by s36srini** [comment]
> considered

You want to be using tf.keras for all sequential models. `from tensorflow.python import keras`, this effectively does the same thing, it's up to you as to whether you want to be explicit or not. What in specific isn't working for you? Are you getting a NoneType returned from the `prune_low_magnitude` wrapper?

**Comment by blakeedwards823** [comment]
`end_step = np.ceil(1.0 * num_train_samples / batch_size).astype(np.int32) * epochs
    # TODO determine how to limit this pruning to retain 90% of the network weights / size
    new_pruning_params = {
        'pruning_schedule': sparsity.PolynomialDecay(initial_sparsity=0.01,
                                                     final_sparsity=0.1,
                                                     begin_step=0,
                                                     end_step=end_step,
                                                     frequency=100)
    }
    new_model = sparsity.prune_low_magnitude(model, **new_pruning_params)
    new_model.compile(
        loss='categorical_crossentropy',
        optimizer='adadelta',
        metrics=['accuracy'])`

This is the code that I am attempting to run in one of my methods. When putting a breakpoint on the compile, where I am getting the error, I see that the new_model is None. 

These are the imports I am using for the model I am training and attempting to prune:
`import os
import tensorflow as tf
from tensorflow.python.keras.layers import MaxPooling2D, Dropout, Dense, Flatten, Activation, Conv2D
from tensorflow.python.keras.models import Model, Sequential, model_from_json
import numpy as np
from HelperUtil import HelpfulFunctions
import datetime
helpful = HelpfulFunctions()`

Is this the incorrect way to go about it?

**Comment by s36srini** [comment]
Your imports look fine to me. Check the type of the model you generated: type(my_model) and lmk what the output is, it should be a tf.keras Model - it cannot be a Sequential Model, you need to convert it. This source code explains why you're getting a NoneType returned: 
```

  if isinstance(to_prune, list):
    return _prune_list(to_prune, **params)
  elif isinstance(to_prune, keras.Model):
    return keras.models.clone_model(
        to_prune, input_tensors=None, clone_function=_add_pruning_wrapper)
  elif isinstance(to_prune, keras.layers.Layer):
    params.update(kwargs)
    return pruning_wrapper.PruneLowMagnitude(to_prune, **params)
```
Take a look at: https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/sparsity/keras/prune.py

Btw, just outta curiosity what do you plan on using pruning for? With this package as it is, you're only going to save footprint size when using a compression utility (e.g gzip). You're also not going to obtain any faster inference time, unless you've enabled some sort of meta parameter on your hardware device.

**Comment by s36srini** [comment]
You can convert it like this: 

```
from keras import layers, models

input_layer = layers.Input(batch_shape=seqmodel.layers[0].input_shape)
prev_layer = input_layer
for layer in seqmodel.layers:
    prev_layer = layer(prev_layer)

funcmodel = models.Model([input_layer], [prev_layer])
```

**Comment by blakeedwards823** [comment]
> Your imports look fine to me. Check the type of the model you generated: type(my_model) and lmk what the output is, it should be a tf.keras Model - it cannot be a Sequential Model, you need to convert it. This source code explains why you're getting a NoneType returned:
> 
> ```
> 
>   if isinstance(to_prune, list):
>     return _prune_list(to_prune, **params)
>   elif isinstance(to_prune, keras.Model):
>     return keras.models.clone_model(
>         to_prune, input_tensors=None, clone_function=_add_pruning_wrapper)
>   elif isinstance(to_prune, keras.layers.Layer):
>     params.update(kwargs)
>     return pruning_wrapper.PruneLowMagnitude(to_prune, **params)
> ```
> 
> Take a look at: https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/sparsity/keras/prune.py
> 
> Btw, just outta curiosity what do you plan on using pruning for? With this package as it is, you're only going to save footprint size when using a compression utility (e.g gzip). You're also not going to obtain any faster inference time, unless you've enabled some sort of meta parameter on your hardware device.

I am pruning a model iteratively and performing other training/tuning in between small pruning steps. I used a custom object to hold my model but returned the sequential keras model. I now understand why it is not working. When printing the type of my model I am seeing:
"Teacher model type:  <class 'TeacherCNN.TeacherModel'>"
Do you know how I can convert this back to a keras model? I am using the keras model in my custom class.

**Comment by alanchiao** [comment]
@blakeedwards823 : it works with pruning a Sequential model. Sequential is an instance keras.Model (see these [docs](https://www.tensorflow.org/api_docs/python/tf/keras/Sequential)).

**Comment by bothrasumit** [comment]
The pruned model type is <class 'NoneType'>
Hence, it does  not work with other calls such as complie or fit or evaluate.
Suggest any solution.

**Comment by alanchiao** [comment]
Closing this issue since Blake has found a resolution [here](https://github.com/tensorflow/model-optimization/issues/12#issuecomment-526350302).
From my guesses, Blake did something effectively like this:
```
  model_wrapper = ModelWrapper()
  model_wrapper.model = model_from_json(...)
  sparsity.prune(model_wrapper)
``
instead of 

```
  sparsity.prune(model_wrapper.model)
```
In other words, the actual Keras model object was not passed. I have filled [an issue]() to add an error message when a Keras model object is not passed. 

@blakeedwards823 : if I'm wrong, please correct me and share the before/after code that demonstrates your resolution. 


@bothrasumit: let's move the discussion to your [issue](https://github.com/tensorflow/model-optimization/issues/164) and see if Blake's problem is the same one you ran into from his response.
